Grid (#2213)
BBC-archive/psammead: packages/components/psammead-grid/README.md

## Description

## Installation

`npm install @bbc/psammead-grid`

## Props

<!-- prettier-ignore -->
| Argument  | Type | Required | Default | Example |
| --------- | ---- | -------- | ------- | ------- |

> | columns | object | yes | N/A | { group1: 6, group3: 8 } |

Reviewer: Example should have all 6 groups

Author: This has been updated

| startOffset | object | no | Sets all values as 1 for each of the groups defined in `columns`  | { group1: 2, group3: 3 } |

> | wrapper | boolean | no | false | `wrapper` |

Reviewer: Descriptions of what each prop does would be nice :)

Author: @dr3 I've added. description below the props table, since adding it in the table was difficult to read. Please let me know if you think the text should be updated.

When we refer to `group3` in this component, we're referring to the breakpoint with a minimum width `GEL_GROUP_3_SCREEN_WIDTH_MIN` & a maximum width `GEL_GROUP_3_SCREEN_WIDTH_MAX`. These values are defined within [`@bbc/gel-foundations/breakpoints`](https://github.com/bbc/psammead/blob/latest/packages/utilities/gel-foundations/src/breakpoints.js).

####<a name="simple-example">Simple example</a>

Here is an example of a `Grid` `wrapper` that has 8 columns for `group4` (from 1008px to 1279px). It has two child `Grid` elements, one which spans 6/8 columns and one which spans 2/8 columns within this breakpoint.

```jsx
import Grid from '@bbc/psammead-grid';

const MyComponent = () => (
  <Grid wrapper columns={{ group4: 8 }}>
    <Grid columns={{ group4: 6 }}>
      <p>Paragraph that spans 6 out of 8 columns through group4</p>
    </Grid>
    <Grid columns={{ group4: 2 }}>
      <p>Paragraph that spans 2 out of 8 columns through group4</p>
    </Grid>
    <Grid columns={{ group4: 2 }}>
       <p>Paragraph that spans 2 out of 8 columns through group4</p>
    </Grid>
    <Grid columns={{ group4: 2 }}>
       <p>Paragraph that spans 2 out of 8 columns through group4</p>
    </Grid>

>   </Grid>

Reviewer 2: As mentioned on the call, think it would be beneficial to have a simple schematic diagram to explain how this would 'look' and what the effect of the different values for `group4` has.

Author: Have added an image below.

);
```

> ####<a name="multiple-groups">Setting values for multiple groups</a>

Reviewer 3: headings need formatted properly

Author: Done

```jsx
import Grid from '@bbc/psammead-grid';

const MyComponent = () => (
  <Grid wrapper columns={{ group3: 6, group4: 8 }}>
    <Grid columns={{ group3: 6, group4: 6 }}>
      <p>Paragraph - for group 3 spans 6/6 columns, for group 4 spans 6/8 columns</p>
    </Grid>
    <Grid columns={{ group3: 6, group4: 2 }}>
      <p>Paragraph - for group 3 spans 6/6 columns, for group 4 spans 2/8 columns</p>
    </Grid>
  </Grid>
);
```

> [Screenshot of this example without GEL Gutters and GEL Margins](./documentation/Screenshot\ example\ without\ gelMargins\ and\ gelGutters.png)

Reviewer 3: Needs fixed

Author: The image has had its syntax updated, but since I've used a relative path, it's pointing to the file location on the `latest` branch, so isn't visible on the PR just yet. Haven't yet found a work-around for this.

####<a name="gutters-margins">Setting standard GEL gutters and GEL Margins</a>

Using `enableGelGutters` and `enableGelMargins` on the wrapper Grid element.

```jsx
import Grid from '@bbc/psammead-grid';

const MyComponent = () => (
	<Grid

> 		wrapper

Reviewer 3: formatting issues

Author: Fixed.

		enableGelGutters
    enableGelMargins
		columns={{ group3: 6, group4: 8 }}
	>
    <Grid columns={{ group3: 6, group4: 6 }}>
      <p>Paragraph - for group 3 spans 6/6 columns, for group 4 spans 6/8 columns</p>
    </Grid>
    <Grid columns={{ group3: 6, group4: 2 }}>
      <p>Paragraph - for group 3 spans 6/6 columns, for group 4 spans 2/8 columns</p>
    </Grid>
  </Grid>

> );

Reviewer 2: What happens if one group is omitted? Does it by default assume the full width? Would be good if we could document the default behaviour.

Reviewer 3: @jamesdonoh 

Does this explain?

> Image of this example when viewed at within group4 (from 1008px to 1279px). When you look at this example at any other breakpoint, each of the four items fill up the full width of the row.

from line 78 of the readme.

    wrapper
    columns={{ group2: 6, group3: 6 }}
  >
		<Grid
		  wrapper
      columns={{ group2: 6, group3: 6 }}
    >
      <Grid columns={{ group2: 6, group3: 3 }}>
        <ExampleImage />
      </Grid>
      <Grid columns={{ group2: 6, group3: 3 }}>
        <ExampleParagraph />
      </Grid>
    </Grid>
    <Grid columns={{ group2: 2, group3: 2 }}>
      <ExampleImage />
    </Grid>
    <Grid columns={{ group2: 4, group3: 4 }}>
      <ExampleParagraph />
    </Grid>
  </Grid>
);
```

[Screenshot of this example with GEL Gutters and GEL Margins enabled](./documentation/Screenshot\ example\ with\ gelMargins\ and\ gelGutters.png)

** Setting GEL Max Widths**
At the page level, you may want to have the GEL Grid's fixed max widths. 
You can enable it with the prop `enableGelMaxWidths`. 

For some pages, we will want to constrain the page to 1008px. You can use `enableGroupFourMaxWidth` for this. 

### When to use this component

This `Grid` component can be used for page-level layouts as well as for layouts of individual components. You can nest Grids within one another for more complex layouts. Please see the Storybook Stories for examples.

### Frequenty Asked Questions

> 

Reviewer 4: I don't know if this is the best place for this question, but I think it'd be good to clarify what happens when you have grid items with a total number of columns that's greater than the wrapper they're in, eg:

```jsx
<Grid wrapper columns={{ group2: 4 }}>
  <Grid columns={{ group2: 3 }}>
    <Child />
  </Grid>
  <Grid columns={{ group2: 2 }}>
    <Child />
  </Grid>
</Grid>
```
I'm guessing the second non-wrapper Grid will move onto a second row and leave the 4th column of the first row empty. Is that correct?

Author: Yes that is correct. I've created this diagram to show what happens in that case.
The first item spans the three columns and the second item spans two on the next row down
<img width="625" alt="four column grid with items three and two" src="https://user-images.githubusercontent.com/3028997/65766642-8734cc00-e123-11e9-96f6-23ad16d627a0.png">

- When should I use the `wrapper` prop?
  - The `wrapper` prop should be used on a parent Grid element. It defines the grid, using the `columns` prop's values to define how many columns to define at each breakpoint. 

>   - It can and should be used whenever you're defining a new grid. [See nested grid example.](#nested-grid). It should be used in any case where you're defining a new nested grid.

Reviewer 4: ```suggestion
  - It can and should be used whenever you're defining a new grid. [See nested grid example](#nested-grid). It should be used in any case where you're defining a new nested grid.
```

Author: Updated the text entirely.

- When should I use the `columns` prop?
  - This should always be defined. For a Grid element that is not a `wrapper`, it's the number of columns it should span.
- When should I use the `startOffset` prop?
  - `startOffset` is an object just like the `columns` prop. You can set it on a Grid child element to start it at a column other than the first one. 
  - If you don't pass it in, the value defaults to 1, the start of the grid.
- Why is there no vertical spacing on the grid?
    - The Grid implementation only has gutters/margins for columns, [according to the GEL specification](https://www.bbc.co.uk/gel/guidelines/grid#grid-sizes).  This is to allow flexibility for a variety of spacing. To add vertical spacing, you should add padding/margin/top/bottom to the contents.

> <!-- ### When not to use this component -->

Reviewer 2: Like, never? 😄

## Description

Psammead Grid is a component that you can use to set out column-based layouts using CSS Grid. It is set up to follow the standard [GEL groups for breakpoints](https://www.bbc.co.uk/gel/guidelines/grid#grid-sizes) and you can toggle on the GEL margins, gutters and fixed widths, in accordance with the GEL grid guidelines if you are using it for page layouts. 

## Installation

`npm install @bbc/psammead-grid`

## Props

<!-- prettier-ignore -->
| Argument | Type | Required | Default | Example |
| -------- | ---- | -------- | ------- | ------- |
| columns | object | yes | N/A | { group1: 6, group2: 6, group3: 6, group4: 8, group5: 20 } |

> | startOffset | object | no | Sets all values as 1 for each of the groups defined in `columns` | { group1: 1, group2: 1, group3: 1, group4: 2, group5: 5 } |

Reviewer 4: I think this has been changed to `gridStartOffset`
```suggestion
| gridStartOffset | object | no | Sets all values as 1 for each of the groups defined in `columns` | { group1: 1, group2: 1, group3: 1, group4: 2, group5: 5 } |
```

Reviewer 3: @AlistairGempf the prop itself hasn't been renamed, it's just been given an alias. I've updated the comment to hopefully make that clearer https://github.com/bbc/psammead/pull/2213/commits/f6d9c05074fdf2d23b61debd5e0e7cc9363d757e

| item | boolean | no | false | `item` |
| enableGelGutters | boolean | no | false | `enableGelGutter` |
| enableGelMargins | boolean | no | false | `enableGelMargins` |
| enableGelMaxWidths | boolean | no | false | `enableGelMaxWidths` |
| enableGroupFourMaxWidth | boolean | no | false | `enableGroupFourMaxWidth` |

- When should I use the `columns` prop?
  - This should always be defined.

>   - For a `<Grid>` element, it's the number of columns the grid has.

Reviewer: ```suggestion
  - For a `<Grid>` element, it's the number of columns the grid has, and the number of columns it should span.
```

#### <a name="simple-example">Simple example</a>

Here is an example of a `<Grid>` that has 8 columns for `group4` (from 1008px to 1279px). It has two child `<Grid item>` elements, one which spans 6/8 columns and one which spans 2/8 columns within this breakpoint.

In this example groups 1, 2, 3, & 5 are omitted from the `columns` prop - so at those breakpoints items fill the whole width of the page. 

```jsx
import Grid from '@bbc/psammead-grid';

const MyComponent = () => (
  <Grid columns={{ group4: 8 }}>
    <Grid item columns={{ group4: 6 }}>
      <p>Item 1 - Paragraph that spans 6 out of 8 columns through group4</p>
    </Grid>
    <Grid item columns={{ group4: 2 }}>
      <p>Item 2 - Paragraph that spans 2 out of 8 columns through group4</p>
    </Grid>
    <Grid item columns={{ group4: 2 }}>
       <p>Item 3 - Paragraph that spans 2 out of 8 columns through group4</p>
    </Grid>
    <Grid item columns={{ group4: 2 }}>
       <p>Item 4 - Paragraph that spans 2 out of 8 columns through group4</p>
    </Grid>
  </Grid>
);
```

> Image of this example

Author: ```suggestion
Image of this example when viewed at within `group4` (from 1008px to 1279px). When you look at this example at any other breakpoint, each of the four items fill up the full width of the row. 
```

- When should I use the `columns` prop?
  - This should always be defined.
  - For a `<Grid>` element, it's the number of columns the grid has, and the number of columns it should span.
  - For a `<Grid item>` element, it's the number of columns it should span.
- When should I use the `item` prop?
  - The `item` prop should be used on a `<Grid>` that is a direct child of another `<Grid>`. Using the `columns` prop's values you can choose how many columns to span at each breakpoint.
  - It can and should be used whenever you're defining a new grid item - something that you want to span a set number of columns at a breakpoint.

```jsx
<Grid columns={{ group5: 6 }}>
  <Grid item columns={{ group5: 1 }}>
    This spans 1 out of 6 columns
  </Grid>
  <Grid item columns={{ group5: 5 }}>
    This spans 5 out of 6 columns
  </Grid>
</Grid>
```

- When should I use the `startOffset` prop?
  - `startOffset` is an object structured just like the `columns` prop. You can set it on a `<Grid item>` element to start it at a column other than the first one.
  - If you don't pass it in, the offset value defaults to 1, the start of the grid.
- Why is there no vertical spacing on the grid?

>   - The Grid implementation only has gutters/margins for columns, [according to the GEL specification](https://www.bbc.co.uk/gel/guidelines/grid#grid-sizes). This is to allow flexibility for a variety of spacing. To add vertical spacing, you should add padding/margin/top/bottom to the contents.

Reviewer 4: This links to grid sizes, and the gutters and margin layout is actually in a section above.
```suggestion
  - The Grid implementation only has gutters/margins for columns, [according to the GEL specification](https://www.bbc.co.uk/gel/guidelines/grid#spacing-and-layout). This is to allow flexibility for a variety of spacing. To add vertical spacing, you should add padding/margin/top/bottom to the contents.
```
